### PR TITLE
[Feature] Allow to use student submissions as an example submission

### LINKED ISSUE #3825: Import Example Submission from Student Submission
### Is your feature request related to a problem?

When preparing example submissions, one possible strategy is looking at random student submissions and copying them as example submissions. Right now, we copy the text/model into the example submission UI and also copy the feedback.

### Describe the solution you'd like

As an Instructor, I want to press and "Import as Example Submission" button after I submitted an assessment and then find the example submission be created with Feedback.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by imstreetspirit**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, editor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [x] Server: I added multiple integration tests (Spring) related to the features (with a high test coverage)
- [x] Server: I added `@PreAuthorize` and check the course groups for all new REST Calls (security)
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.
- [x] Client: I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/).
- [x] Client: I added multiple integration tests (Jest) related to the features (with a high test coverage)
- [x] Client: I added `authorities` to all new routes and check the course groups for displaying navigation elements (links, buttons)
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I added multiple screenshots/screencasts of my UI changes
- [x] Client: I translated all newly inserted strings into English and German.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).

### Motivation and Context

Fixes #3825 

Example submission is useful to train tutors before the assessment period. Instructor sometimes looks at the random student submission and decides to use this submission as an example submission. For that, instructor has to copy the text/model element and the assessment of the submission. This PR introduces new “Import as Example Submission” button in the assessment page. With this feature, instructor able to copy the student submission with its assessment, then instructor can find this newly created example submission in the board.

### Description

- Example submission is available for Modeling and Text exercise. `Import Example Submission` button is added in `assessment-header-component,` the corresponding event functions are defined in `modeling-assessment-editor.component` and `text-submission-assessment.component`

- New route for import API is defined in `ExampleSubmissionResource`, and the corresponding api call service is defined in `example-submission.service` 

- `importStudentSubmissionAsExampleSubmission()` is introduced to use corresponding `copySubmission `functions depending exercise type. 

While working on this implementation, I noticed some problems.  Since it affects the functionality of this feature, so this PR also covers bugfixing for them:

**Bugfix**:  If the text exercise has already example submission, there is Internal Server Error when we want to Import the Text Exercise. 

- This happens because some text-block variables(e.g `addedDistance`, `cluster`, `positionInCluster`) can be null.  Because, `addedDistance` is only needed for feedback suggestions, and copied text-blocks are not part of the `cluster`. When we copy the text blocks these null variables cause `NullPointerException`. So, the corresponding null checks are added for `copyTextBlocks `in `TextExerciseImportService`

**Bugfix:** After copying the submission with its assessment, the feedback references are not matched with text-block ids which is copied from original submission.

- After creating new text block, submission, result, feedback entries by referencing the original submission, the feedback references still match with original text block ids. So, I introduced new function called `updateFeedbackReferencesWithNewTextBlockIds()` in `TextExerciseImportService`. So, the assessment of new submission will be shown properly after importing the text exercise.

Also refactored the client testBed configuration to speed-up tests for (`text-submission-assessment.component.spec.ts` and `example-submission.service.spec.ts` ) (because it took so long while I was working on tests)

### Steps for Testing

1. Login Artemis
2. Navigate to Course Management
3. Find/create modeling and text exercises (Instructor)
4. The exercise should have at least one student submission to utest the functionality, if your exercise already have one skip this step, otherwise submit new submisison as Student
5. As an Instructor, assess the student submission, save the assessment then you should see “Import as Example Submission” button.
6. Click that, then go to Example Submission page for that exercise.
7. The example submission should be clisted in there, open the submission, confirm that it is copied properly with its assessment.

**Testing Steps For Bugfix:** 

1. Login Artemis
2. Navigate to Course Management
3. Find/create text exercises (Instructor)
4. Text exercise must have at least one example submission, do not forget to assess the example submission.
5. Then go back to Exercises page, Click “Import Text Exercise” button, choose the exercise that you created in previous steps.
6. Change the name if you want (you can find the imported exercises easily afterwards), click “Import”. 
7. Confirm that there is no Internal Server Error.
8. Go to Example Submissions page for the imported exercise.
9. Check the example submission, , confirm that it is copied properly with its assessment.

### Review Progress

- Code Review
  - [x] Review 1
  - [x] Review 2
- Manual Tests
  - Test for Import as Example Submission button functionality
      - [x] Test 1 
      - [x] Test 2  
  - Test for Import Text Exercise with Example Submissions (Bugfix)
      - [x] Test 1 (From Current Course to Another Course)
      - [x] Test 2 (From Current Course to Same Course)

### Test Coverage

*only covered new functions
ExampleSubmissionService.java 89%
TextExerciseImportService.java 100%
ExampleSubmissionResource.java 86%
example-submission.service.spec.ts 100%
text-submission-assessment.component.spec.ts 68%
modeling-assessment-editor.component.spec.ts 75%

### Screenshots

Import as Example Submission Button in Assessment Page

Example Submission board for the respective exercise

### CODE REVIEW

**Review by MerlinMehmed — COMMENTED** [review]
I tested on TS1, the bug is fixed and import of a text exercise works without an internal server error and the example submission is copied correctly.
I have a question regarding the import of example submission. When assessing student's submission I added additional feedback shown in the image below
Screenshot from 2021-08-24 08-35-48
and this feedback was not added in the example submission (see below)
Screenshot from 2021-08-24 08-36-19
I am not completely sure what is the use case of the feedback but with it, the assessment is graded for 6/10 points and the example submission has 1/10 points just because the feedback is missing. Am I using it wrong and is it correct that this part is not copied?

**Comment by imstreetspirit** [comment]
> I have a question regarding the import of example submission. When assessing student's submission I added additional feedback shown in the image below

hi @MerlinMehmed ! thanks for testing it. so you override the assessment right? If you add the feedback after you already used this submission as an example submission, it won't be displayed in the example submission. Because, when you click this Import button, it copies the current submission and creates new one, so the newly created example submission is not associated student submission directly.

> I am not completely sure what is the use case of the feedback but with it, the assessment is graded for 6/10 points and the example submission has 1/10 points just because the feedback is missing. Am I using it wrong and is it correct that this part is not copied?

I thought that instructor may want to override the assessment after importing the student submission. Import button will be activated, when the assessment has been submitted once. So that, he/she can create different combinations of this assessment to use these example submission in training process of assessment.  Is this more clear from your side?

**Review by MerlinMehmed — APPROVED** [review]
Thank you for the detailed description @imstreetspirit. It's clear now. Having said that I think that the import example submission functionality works as expected.
I also had a look at the code and it looks good to me.

File: src/main/webapp/i18n/de/assessment.json [hunk 1]
```diff
@@ -28,7 +28,8 @@
             "button": {
                 "overrideAssessment": "Bewertung überschreiben",
                 "nextSubmission": "Nächste Abgabe bewerten",
-                "resolveConflict": "Konflikte behandeln"
+                "resolveConflict": "Konflikte behandeln",
+                "importExampleSubmission": "Import als Beispielabgabe"
```
Comment (MDK5) [inline src/main/webapp/i18n/de/assessment.json:32]:
```suggestion
                "importExampleSubmission": "Als Beispielabgabe importieren"
```
to keep in like with the "object verb" structure of the other strings

**Review by MDK5 — APPROVED** [review]
Reviewed code during testing session, changes look good to me, just found a few small suggestions:

**Review by ndricimrr — DISMISSED** [review]
I tested on TS3 with Text Exercises for both the test cases: 
- Test for Import as Example Submission button functionality
- Test for Import Text Exercise with Example Submissions (Bugfix)
Works fine for me, didn't encounter any issues, nor internal server error logs 👍

File: src/test/javascript/spec/component/text-submission-assessment/text-submission-assessment.component.spec.ts [hunk 2]
```diff
@@ -121,30 +132,30 @@ describe('TextSubmissionAssessmentComponent', () => {
...
-                ArtemisTestModule,
-                ArtemisSharedModule,
-                ArtemisAssessmentSharedModule,
-                AssessmentInstructionsModule,
-                TranslateModule.forRoot(),
-                ArtemisConfirmIconModule,
-                TextSharedModule,
-                RouterTestingModule,
-                ArtemisGradingInstructionLinkIconModule,
-            ],
```
Comment (Olimbia23) [inline src/test/javascript/spec/component/text-submission-assessment/text-submission-assessment.component.spec.ts:136]: Improving test code speed. Nice 👍

**Review by Olimbia23 — APPROVED** [review]
Checked the code. Nice work @imstreetspirit 💯 
I like that you speed up the tests as well! Approve 👍

File: src/test/javascript/spec/component/text-submission-assessment/text-submission-assessment.component.spec.ts (hunk 2 above)
Comment (imstreetspirit) [reply src/test/javascript/spec/component/text-submission-assessment/text-submission-assessment.component.spec.ts:136]: haha, working with the old suit was like a hell, i had to wait over 15min to run only one test case.

File: src/test/java/de/tum/in/www1/artemis/ExampleSubmissionIntegrationTest.java [hunk 3]
```diff
@@ -243,4 +250,73 @@ public void createExampleTextAssessment_wrongExerciseId() throws Exception {
...
+        textBlock.setStartIndex(0);
+        textBlock.setEndIndex(14);
+        database.addAndSaveTextBlocksToTextSubmission(Set.of(textBlock), (TextSubmission) submission);
+
+        database.addResultToSubmission(submission, AssessmentType.MANUAL);
+
+        // add one feedback for the created text block
+        List<TextBlock> textBlocks = new ArrayList<>(((TextSubmission) submission).getBlocks());
+        Feedback feedback = new Feedback();
+        if (!textBlocks.isEmpty()) {
```
Comment (julian-christl) [inline src/test/java/de/tum/in/www1/artemis/ExampleSubmissionIntegrationTest.java:275]: Should we fail the test if the textBlocks are empty? Since at this point, we kinda expect that there are some since we added them. If there are none, this should be an exceptional behaviour that must be fixed, shouldn't it?

File: src/test/java/de/tum/in/www1/artemis/ExampleSubmissionIntegrationTest.java [hunk 4]
```diff
@@ -243,4 +250,73 @@ public void createExampleTextAssessment_wrongExerciseId() throws Exception {
...
+            feedback.setReference(textBlocks.get(0).getId());
+        }
+
+        database.addFeedbackToResult(feedback, submission.getLatestResult());
+
+        ExampleSubmission exampleSubmission = importExampleSubmission(HttpStatus.OK, submission.getId(), textExercise.getId());
+        List<TextBlock> copiedTextBlocks = new ArrayList<>(((TextSubmission) exampleSubmission.getSubmission()).getBlocks());
+        assertThat(exampleSubmission.getId()).isNotNull();
+        assertThat(((TextSubmission) exampleSubmission.getSubmission()).getText()).isEqualTo(((TextSubmission) submission).getText());
+        assertThat(exampleSubmission.getSubmission().getLatestResult().getFeedbacks().size()).isEqualTo(1);
```
Comment (julian-christl) [inline src/test/java/de/tum/in/www1/artemis/ExampleSubmissionIntegrationTest.java:286]: Would an equal check on the feedback itself make more sense? => assertThat getFeedbacks() returns an Set containing that feedback.

File: src/test/java/de/tum/in/www1/artemis/ExampleSubmissionIntegrationTest.java [hunk 5]
```diff
@@ -243,4 +250,73 @@ public void createExampleTextAssessment_wrongExerciseId() throws Exception {
...
+        assertThat(copiedTextBlocks.size()).isEqualTo(1);
```
Comment (julian-christl) [inline src/test/java/de/tum/in/www1/artemis/ExampleSubmissionIntegrationTest.java:287]: Same here regarding equality check.

File: src/test/java/de/tum/in/www1/artemis/ExampleSubmissionIntegrationTest.java [hunk 6]
```diff
@@ -243,4 +250,73 @@ public void createExampleTextAssessment_wrongExerciseId() throws Exception {
...
+
+    @Test
+    @WithMockUser(value = "instructor1", roles = "INSTRUCTOR")
+    public void importExampleSubmissionWithModelingSubmission() throws Exception {
+        Submission submission = ModelFactory.generateModelingSubmission(validModel, true);
+        submission = database.addModelingSubmission(modelingExercise, (ModelingSubmission) submission, "student1");
+        database.addResultToSubmission(submission, AssessmentType.MANUAL);
+
+        ExampleSubmission exampleSubmission = importExampleSubmission(HttpStatus.OK, submission.getId(), modelingExercise.getId());
+        assertThat(exampleSubmission.getId()).isNotNull();
```
Comment (julian-christl) [inline src/test/java/de/tum/in/www1/artemis/ExampleSubmissionIntegrationTest.java:299]: Please add a check to assert equality between the modelling strings as well (as with the text in the text exercise)

File: src/test/javascript/spec/component/assessment-shared/assessment-header.component.spec.ts [hunk 7]
```diff
@@ -1,30 +1,42 @@
...
-    beforeEach(async(() => {
+    beforeEach(() => {
         TestBed.configureTestingModule({
-            imports: [ArtemisTestModule, ArtemisSharedModule, RouterTestingModule],
-            declarations: [AssessmentHeaderComponent, MockComponent(AssessmentWarningComponent)],
+            imports: [ArtemisTestModule, RouterTestingModule],
+            declarations: [
+                AssessmentHeaderComponent,
+                MockComponent(AssessmentWarningComponent),
+                AlertComponent,
```
Comment (julian-christl) [inline src/test/javascript/spec/component/assessment-shared/assessment-header.component.spec.ts:33]: Can this component be mocked as well?

File: src/test/javascript/spec/component/modeling-assessment-editor/modeling-assessment-editor.component.spec.ts [hunk 8]
```diff
@@ -469,4 +473,23 @@ describe('ModelingAssessmentEditorComponent', () => {
...
+    it('should invoke import example submission', () => {
+        const course = new Course();
+        component.modelingExercise = new ModelingExercise(UMLDiagramType.ClassDiagram, course, undefined);
+        component.modelingExercise.id = 1;
+        component.submission = {
+            id: 2,
+            submitted: true,
+            type: 'MANUAL',
+            text: 'Test\n\nTest\n\nTest',
+        } as unknown as ModelingSubmission;
```
Comment (julian-christl) [inline src/test/javascript/spec/component/modeling-assessment-editor/modeling-assessment-editor.component.spec.ts:485]: Could you explain to me what this does? Why not just `as ModelingSubmission`?

File: src/test/javascript/spec/component/text-submission-assessment/text-submission-assessment.component.spec.ts [hunk 9]
```diff
@@ -160,25 +171,30 @@ describe('TextSubmissionAssessmentComponent', () => {
...
         fixture.detectChanges();
     });
 
+    afterEach(() => {
+        sinon.restore();
+    });
+
     it('should create', () => {
-        expect(component).toBeTruthy();
+        expect(component).to.be.ok;
```
Comment (julian-christl) [inline src/test/javascript/spec/component/text-submission-assessment/text-submission-assessment.component.spec.ts:184]: When looking up the definition I saw that they recommend the explicit value: https://www.chaijs.com/api/bdd/#method_ok
So I assume for this use case it would be `.to.not.be.undefined`.

Other question: What's the reason behind this change?

File: src/test/javascript/spec/component/text-submission-assessment/text-submission-assessment.component.spec.ts [hunk 10]
```diff
@@ -207,33 +223,28 @@ describe('TextSubmissionAssessmentComponent', () => {
...
 
-        spyOn(textAssessmentService, 'save').and.returnValue(
-            of(
-                new HttpResponse({
-                    body: result,
-                }),
-            ),
-        );
+        const saveStub = sinon.stub(textAssessmentService, 'save');
+        saveStub.returns(of(new HttpResponse({ body: result })));
```
Comment (julian-christl) [inline src/test/javascript/spec/component/text-submission-assessment/text-submission-assessment.component.spec.ts:227]: Nice simplification!

**Review by julian-christl — CHANGES_REQUESTED** [review]
Great changes and great feature implementation.
I have some questions and a few suggestions.

File: src/test/javascript/spec/component/text-submission-assessment/text-submission-assessment.component.spec.ts (hunk 9 above)
Comment (imstreetspirit) [reply src/test/javascript/spec/component/text-submission-assessment/text-submission-assessment.component.spec.ts:184]: this test suit is very old, we are not using Jasmine or we try to avoid them. As you can see, I imported sinon and chai, and I decided to adapt the whole test suit in [recommended way ](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/client-tests/)

File: src/main/java/de/tum/in/www1/artemis/service/TextExerciseImportService.java [hunk 11]
```diff
@@ -132,7 +136,43 @@ TextSubmission copySubmission(final Submission originalSubmission) {
...
+        Result result = newSubmission.getLatestResult();
+        List<Feedback> feedbackList = result.getFeedbacks();
+        Set<TextBlock> newSubmissionTextBlocks = newSubmission.getBlocks();
+
+        // first collect original text blocks as <startIndex, TextBlock> map, startIndex will help to match newly created text block with original text block
+        Map<Integer, TextBlock> originalTextBlockMap = originalTextBlocks.stream().collect(Collectors.toMap(TextBlock::getStartIndex, Function.identity()));
+
+        Map<String, String> textBlockIdPair = new HashMap<>();
+
+        // collect <original text block id, new text block id> pair, it will help to find the feedback which has old reference
```
[the remark this answers is not in the document]
Comment (imstreetspirit) [reply src/main/java/de/tum/in/www1/artemis/service/TextExerciseImportService.java:162]: maybe, I did not quite understand but how can the textblocks have same start index? Think one paragraph and each sentence in this paragraph, each sentence starts from different index. or am I missing something in here?

File: src/main/java/de/tum/in/www1/artemis/web/rest/ExampleSubmissionResource.java [hunk 12]
```diff
@@ -145,4 +151,25 @@ public ExampleSubmissionResource(ExampleSubmissionService exampleSubmissionServi
...
     }
+
+    /**
+     * POST exercises/:exerciseId/example-submissions/import : Import exampleSubmission.
+     *
+     * @param exerciseId        the id of the corresponding exercise
+     * @param submissionId        the submission id to be imported as an example submission
+     * @return the ResponseEntity with status 200 (OK) and the Result as its body, or with status 4xx if the request is invalid
+     */
+    @PostMapping("exercises/{exerciseId}/example-submissions/import")
```
[the remark this answers is not in the document]
Comment (imstreetspirit) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ExampleSubmissionResource.java:162]: one reminder, we are not using CamelCase for mapping. hmm, could be. in my first try I was getting whole submission object from the client side. after deciding that this is not necessary I just changed it as getting only submissionId.However submissionId can be extracted from path.

File: src/main/webapp/app/assessment/assessment-header/assessment-header.component.html [hunk 13]
```diff
@@ -123,6 +123,15 @@ <h3 class="top-container">
             <fa-icon *ngIf="nextSubmissionBusy" [icon]="'spinner'" [spin]="true">&nbsp;</fa-icon>
             <span jhiTranslate="artemisApp.assessment.button.nextSubmission"> Next Submission</span>
         </button>
+        <button
+            class="btn ms-2 btn-primary"
+            id="importExampleSubmission"
+            *ngIf="result?.completionDate && isAtLeastInstructor && (exercise!.type === exerciseType.MODELING || exercise!.type === exerciseType.TEXT)"
+            (click)="importExampleSubmission.emit()"
+        >
+            <fa-icon *ngIf="submitBusy" [icon]="'spinner'" [spin]="true"></fa-icon>
```
[the remark this answers is not in the document]
Comment (imstreetspirit) [reply src/main/webapp/app/assessment/assessment-header/assessment-header.component.html:132]: you can check the icon directive usage in Artemis.

File: src/main/webapp/app/assessment/assessment-header/assessment-header.component.html [hunk 14]
```diff
@@ -123,6 +123,15 @@ <h3 class="top-container">
...
+            <span jhiTranslate="artemisApp.assessment.button.importExampleSubmission"> Import as Example Submission</span>
```
[the remark this answers is not in the document]
Comment (imstreetspirit) [reply src/main/webapp/app/assessment/assessment-header/assessment-header.component.html:133]: I can remove whitespace To be more clear I think putting default value makes sense to me, there are so many examples in Artemis. I did not see the usage you mentioned in Artemis so far.

File: src/main/webapp/app/exercises/text/assess/text-submission-assessment.component.ts [hunk 15]
```diff
@@ -457,4 +460,16 @@ export class TextSubmissionAssessmentComponent extends TextAssessmentBaseCompone
...
+     * Invokes exampleSubmissionService when importExampleSubmission is emitted in assessment-layout
+     */
+    importStudentSubmissionAsExampleSubmission(): void {
+        if (this.submission && this.exercise) {
+            this.exampleSubmissionService.import(this.submission, this.exercise.id!).subscribe(
+                () => this.jhiAlertService.success('artemisApp.exampleSubmission.submitSuccessful'),
+                (error: HttpErrorResponse) => onError(this.jhiAlertService, error),
+            );
+        }
+    }
```
[the remark this answers is not in the document]
Comment (imstreetspirit) [reply src/main/webapp/app/exercises/text/assess/text-submission-assessment.component.ts:474]: it can be refactored later on, but if you can check the buttons in assessment-header.component. The listeners are always defined separately for each exercise type submission component. For this reason, I need to implement same structure for this Import button I introduced newly.

File: src/test/java/de/tum/in/www1/artemis/ExampleSubmissionIntegrationTest.java [hunk 16]
```diff
@@ -243,4 +250,73 @@ public void createExampleTextAssessment_wrongExerciseId() throws Exception {
         assertThat(exampleSubmissionRepo.findBySubmissionId(randomId)).isEmpty();
     }
 
+    private ExampleSubmission importExampleSubmission(HttpStatus expectedStatus, Long submissionId, Long exerciseId) throws Exception {
```
[the remark this answers is not in the document]
Comment (imstreetspirit) [reply src/test/java/de/tum/in/www1/artemis/ExampleSubmissionIntegrationTest.java:253]: maybe I missed something in here, which parameter does not math with the call?

File: src/test/javascript/spec/component/assessment-shared/assessment-header.component.spec.ts (hunk 7 above)
Comment (imstreetspirit) [reply src/test/javascript/spec/component/assessment-shared/assessment-header.component.spec.ts:33]: I tried to mock it before, but for some reason it did not work, so I just put the component itself.

File: src/test/javascript/spec/component/modeling-assessment-editor/modeling-assessment-editor.component.spec.ts (hunk 8 above)
Comment (imstreetspirit) [reply src/test/javascript/spec/component/modeling-assessment-editor/modeling-assessment-editor.component.spec.ts:485]: I used similar initialization what we used in the lines above. normally, it sets an unknown type to other type. But I think it can be defined without unknown.

File: src/test/javascript/spec/component/text-submission-assessment/text-submission-assessment.component.spec.ts (hunk 9 above)
Comment (imstreetspirit) [reply src/test/javascript/spec/component/text-submission-assessment/text-submission-assessment.component.spec.ts:184]: this test suit is very old, we are not using Jasmine or we try to avoid them. As you can see, I imported sinon and chai, and I decided to adapt the whole test suit in [recommended way ](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/client-tests/)

File: src/test/java/de/tum/in/www1/artemis/ExampleSubmissionIntegrationTest.java (hunk 4 above)
Comment (imstreetspirit) [reply src/test/java/de/tum/in/www1/artemis/ExampleSubmissionIntegrationTest.java:286]: do you mean comparing the feedback I created with the feedback in the example submission?

File: src/test/java/de/tum/in/www1/artemis/ExampleSubmissionIntegrationTest.java (hunk 5 above)
Comment (imstreetspirit) [reply src/test/java/de/tum/in/www1/artemis/ExampleSubmissionIntegrationTest.java:287]: same as above..

File: src/main/java/de/tum/in/www1/artemis/web/rest/ExampleSubmissionResource.java [hunk 17]
```diff
@@ -145,4 +151,25 @@ public ExampleSubmissionResource(ExampleSubmissionService exampleSubmissionServi
...
+    @PreAuthorize("hasRole('INSTRUCTOR')")
+    public ResponseEntity<ExampleSubmission> importExampleSubmission(@PathVariable Long exerciseId, @RequestBody Long submissionId) {
+        log.debug("REST request to import ExampleSubmission : {}", submissionId);
+
+        Exercise exercise = exerciseRepository.findByIdElseThrow(exerciseId);
+
+        authCheckService.checkHasAtLeastRoleForExerciseElseThrow(Role.INSTRUCTOR, exercise, null);
+
```
[the remark this answers is not in the document]
Comment (imstreetspirit) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ExampleSubmissionResource.java:170]: that edge case seems unlikely, but yeah can be checked

File: src/main/webapp/app/exercises/shared/example-submission/example-submission.service.ts [hunk 18]
```diff
@@ -57,6 +58,19 @@ export class ExampleSubmissionService {
         return this.http.delete<void>(`api/example-submissions/${exampleSubmissionId}`, { observe: 'response' });
     }
 
+    /**
+     * Imports an example submission
+     * @param submission the submission to be imported as an example submission
+     * @param exerciseId the id of the corresponding exercise
+     */
+    import(submission: Submission, exerciseId: number): Observable<EntityResponseType> {
```
[the remark this answers is not in the document]
Comment (imstreetspirit) [reply src/main/webapp/app/exercises/shared/example-submission/example-submission.service.ts:66]: maybe I can get the submissionId directly after changing the mapping

File: src/test/java/de/tum/in/www1/artemis/ExampleSubmissionIntegrationTest.java (hunk 3 above)
Comment (imstreetspirit) [reply src/test/java/de/tum/in/www1/artemis/ExampleSubmissionIntegrationTest.java:275]: hmm did not think about it, yeah could be.

File: src/main/java/de/tum/in/www1/artemis/service/TextExerciseImportService.java [hunk 19]
```diff
@@ -151,7 +151,7 @@ TextSubmission copySubmission(final Submission originalSubmission) {
      */
     private void updateFeedbackReferencesWithNewTextBlockIds(Set<TextBlock> originalTextBlocks, TextSubmission newSubmission) {
         Result result = newSubmission.getLatestResult();
```
Comment (julian-christl) [inline src/main/java/de/tum/in/www1/artemis/service/TextExerciseImportService.java:153]: I think you missed this one when adding the `new` prefix :D

File: src/main/java/de/tum/in/www1/artemis/service/TextExerciseImportService.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (julian-christl) [reply src/main/java/de/tum/in/www1/artemis/service/TextExerciseImportService.java:162]: Ah. No, you are correct. Currently, that's the only option, otherwise, errors get thrown. Then this should not be an issue.

File: src/main/java/de/tum/in/www1/artemis/web/rest/ExampleSubmissionResource.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (julian-christl) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ExampleSubmissionResource.java]: Good point with the CamelCase!

File: src/main/webapp/app/assessment/assessment-header/assessment-header.component.html (hunk 13 above)
[the remark this answers is not in the document]
Comment (julian-christl) [reply src/main/webapp/app/assessment/assessment-header/assessment-header.component.html:132]: Thanks. I'll do!

File: src/main/webapp/app/assessment/assessment-header/assessment-header.component.html (hunk 14 above)
[the remark this answers is not in the document]
Comment (julian-christl) [reply src/main/webapp/app/assessment/assessment-header/assessment-header.component.html]: Two things:
1. I searched for it but didn't find it as well. Not sure what I remembered. Sorry about that.
2. I agree that we currently have multiple ways to write it spread over the whole system. With the jhiTranslate attribute, as an Angular input with the translation pipeline, with whitespace and without. Let's discuss this next meeting maybe what approach to take to make it consistent.

=> convert the whitespace though to the HTML character. This makes sure that the whitespace character can't be use for a linebreak
=> `&nbsp;`

File: src/main/webapp/app/exercises/text/assess/text-submission-assessment.component.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (julian-christl) [reply src/main/webapp/app/exercises/text/assess/text-submission-assessment.component.ts]: I see so there would be major refactoring in the component and templates themselves necessary 👍

File: src/test/java/de/tum/in/www1/artemis/ExampleSubmissionIntegrationTest.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (julian-christl) [reply src/test/java/de/tum/in/www1/artemis/ExampleSubmissionIntegrationTest.java:253]: I think my statement was ambiguous:
You call the postWithResponseBody in this method with your parameters in the following order: `exerciseId`, `submissionId`, `expectedStatus`.
But you receive the parameters with `expectedStatus`, `submissionId`, `exerciseId`
=> The other way around. I think it makes sense to stay consistent so it's more intuitive to read. What do you think?

File: src/test/java/de/tum/in/www1/artemis/ExampleSubmissionIntegrationTest.java (hunk 4 above)
Comment (julian-christl) [reply src/test/java/de/tum/in/www1/artemis/ExampleSubmissionIntegrationTest.java]: I meant it a different way but it makes sense the way you changed it 👍

File: src/test/javascript/spec/component/text-submission-assessment/text-submission-assessment.component.spec.ts (hunk 9 above)
Comment (julian-christl) [reply src/test/javascript/spec/component/text-submission-assessment/text-submission-assessment.component.spec.ts:184]: Thanks for the explanation makes sense. Then I'll ask the person that wrote the documentation.

File: src/test/java/de/tum/in/www1/artemis/ExampleSubmissionIntegrationTest.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (imstreetspirit) [reply src/test/java/de/tum/in/www1/artemis/ExampleSubmissionIntegrationTest.java:253]: ohhh ,I see now... hmm, why not

File: src/main/java/de/tum/in/www1/artemis/web/rest/ExampleSubmissionResource.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (julian-christl) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ExampleSubmissionResource.java:170]: While it seems unlikely it is a security issue because, in theory, one could import that way a student submission from another course even into this exercise which would allow this person to look at submissions, potentially even exam submissions, that there person doesn't have access to.

[the remark this answers is not in the document]
Comment (imstreetspirit) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ExampleSubmissionResource.java:170]: hmm, cool. I'm convinced. 😄 need to check submission repo what I can do about that. cause now I have only submissionId and exerciseId,

[the remark this answers is not in the document]
Comment (julian-christl) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ExampleSubmissionResource.java:170]: I think you can check whether `Submission.of(submissionId).participation.exercise.id == exerciseId`.
But I might be too premature on that.

**Review by julian-christl — DISMISSED** [review]
Thank you for the changes and answers to my questions. The code looks good to me now.

**Review by julian-christl — APPROVED** [review]
Reapproval

**Review by Olimbia23 — APPROVED** [review]
Rechecked the code. Reapprove 👍

**Review by ndricimrr — APPROVED** [review]
Tested again on TS1 and worked fine. Re-Approve

**Review by MerlinMehmed — APPROVED** [review]
Reapprove. Tested on TS1 and worked as expected.

**Comment by Olimbia23** [comment]
I also tested the functionality for import example submission for text and modelling exercise in TS1
It works as expected 🎉 @imstreetspirit
